src/Processors/Transforms/DeduplicationTokenTransforms.cpp

```diff
@@ -0,0 +1,241 @@
+#include <Processors/Transforms/DeduplicationTokenTransforms.h>
+
+#include <IO/WriteHelpers.h>
+
+#include <Common/logger_useful.h>
+#include <Common/Exception.h>
+#include <Common/SipHash.h>
+
+
+#include <fmt/core.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+void RestoreChunkInfosTransform::transform(Chunk & chunk)
+{
+    chunk.getChunkInfos().append(chunk_infos.clone());
+}
+
+namespace DeduplicationToken
+{
+
+String TokenInfo::getToken() const
+{
+    if (!isDefined())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is not defined, stage {}, token {}", stage, debugToken());
+
+    return getTokenImpl();
+}
+
+String TokenInfo::getTokenImpl() const
+{
+    String result;
+    result.reserve(getTotalSize());
+
+    for (const auto & part : parts)
+    {
+        if (!result.empty())
+            result.append(":");
+        result.append(part);
+    }
+
+    return result;
+}
+
+String TokenInfo::debugToken() const
+{
+    return getTokenImpl();
+}
+
+void TokenInfo::addChunkHash(String part)
+{
+    if (stage == UNDEFINED)
+        stage = DEFINE_SOURCE_WITH_HASHES;
+
+    if (stage != DEFINE_SOURCE_WITH_HASHES)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is in wrong stage {}, token {}", stage, debugToken());
+
+    addTokenPart(std::move(part));
+}
+
+void TokenInfo::defineSourceWithChunkHashes()
+{
+    if (stage == UNDEFINED && empty())
+        stage = DEFINE_SOURCE_WITH_HASHES;
+
+    if (stage != DEFINE_SOURCE_WITH_HASHES)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is in wrong stage {}, token {}", stage, debugToken());
+
+    stage = DEFINED;
+}
+
+void TokenInfo::setUserToken(const String & token)
+{
+    if (stage == UNDEFINED)
+        stage = DEFINE_SOURCE_USER_TOKEN;
+
+    if (stage != DEFINE_SOURCE_USER_TOKEN)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is in wrong stage {}, token {}", stage, debugToken());
+
+    addTokenPart(fmt::format("user-token-{}", token));
+}
+
+void TokenInfo::defineSourceWithUserToken(size_t block_number)
+{
+    if (stage != DEFINE_SOURCE_USER_TOKEN)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is in wrong stage {}, token {}", stage, debugToken());
+
+    addTokenPart(fmt::format("source-number-{}", block_number));
+
+    stage = DEFINED;
+}
+
+void TokenInfo::setViewID(const String & id)
+{
+    if (stage == DEFINED)
+        stage = DEFINE_VIEW;
+
+    if (stage != DEFINE_VIEW)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is in wrong stage {}, token {}", stage, debugToken());
+
+    addTokenPart(fmt::format("view-id-{}", id));
+}
+
+void TokenInfo::defineViewID(size_t block_number)
```

Comment:
Should this method be called `setViewBlockNumber` instead? I see you probably called it `define` because after this method stage becomes `DEFINED`, but it is kind of strange `set` and `define` difference, which is misleading and you start to wonder why they differ, while in fact they do not, the only difference is the next strage, while the user should not really know about stages, as they are internal to this class